src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterDocumentRewriter.cs

```diff
@@ -0,0 +1,696 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.IntroduceVariable
+{
+    internal abstract partial class AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax>
+    {
+        private class IntroduceParameterDocumentRewriter
+        {
+            public AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax> Service { get; }
+            public Document OriginalDocument { get; }
+            public TExpressionSyntax Expression { get; }
+            public IMethodSymbol MethodSymbol { get; }
+            public SyntaxNode ContainerMethod { get; }
+            public IntroduceParameterCodeActionKind ActionKind { get; }
+            public bool AllOccurrences { get; }
+
+            public IntroduceParameterDocumentRewriter(AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax> service,
+                Document originalDocument,
+                TExpressionSyntax expression,
+                IMethodSymbol methodSymbol,
+                SyntaxNode containingMethod,
+                IntroduceParameterCodeActionKind selectedCodeAction,
+                bool allOccurrences)
+            {
+                Service = service;
+                OriginalDocument = originalDocument;
+                Expression = expression;
+                MethodSymbol = methodSymbol;
+                ContainerMethod = containingMethod;
+                ActionKind = selectedCodeAction;
+                AllOccurrences = allOccurrences;
+            }
+
+            /// <summary>
+            /// Ties the identifiers within the expression back to their associated parameter.
+            /// </summary>
+            public async Task<Dictionary<TIdentifierNameSyntax, IParameterSymbol>> MapExpressionToParametersAsync(CancellationToken cancellationToken)
```

Comment:
do all these methods need to be public?

Reply:
No besides the one called in the abstract class, changed.